server/notifications.go

```diff
@@ -29,22 +47,48 @@ func (ah *ActivityHandler) handleCreatedActivityNotification(msg *clientmodels.M
 			continue
 		} else if err != nil {
 			ah.plugin.GetAPI().LogWarn("Failed to map Teams user to Mattermost user", "teams_user_id", member.UserID, "error", err)
+			ah.plugin.metricsService.ObserveNotification(isGroupChat, hasFilesUnknown, metrics.DiscardedReasonInternalError)
 			continue
 		}
 
 		if !ah.plugin.getNotificationPreference(mattermostUserID) {
+			ah.plugin.GetAPI().LogInfo(
+				"Skipping notification for chat member who disabled notifications",
+				"user_id", mattermostUserID,
+				"teams_user_id", member.UserID,
+				"chat_id", chat.ID,
+				"message_id", msg.ID,
+			)
+			ah.plugin.metricsService.ObserveNotification(isGroupChat, hasFilesUnknown, metrics.DiscardedReasonUserDisabledNotifications)
+			continue
+		}
+
+		// Don't notify users active in Teams.
+		if userPresenceIsActive(presences[member.UserID]) {
+			ah.plugin.GetAPI().LogInfo(
+				"Skipping notification for chat member present in Teams",
+				"user_id", mattermostUserID,
+				"teams_user_id", member.UserID,
+				"chat_id", chat.ID,
+				"message_id", msg.ID,
+				"activity", presences[member.UserID].Activity,
+				"availability", presences[member.UserID].Availability,
+			)
+			ah.plugin.metricsService.ObserveNotification(isGroupChat, hasFilesUnknown, metrics.DiscardedReasonUserActiveInTeams)
 			continue
 		}
 
 		channel, err := ah.plugin.apiClient.Channel.GetDirect(mattermostUserID, ah.plugin.botUserID)
 		if err != nil {
 			ah.plugin.GetAPI().LogWarn("Failed to get bot DM channel with user", "user_id", mattermostUserID, "teams_user_id", member.UserID, "error", err)
+			ah.plugin.metricsService.ObserveNotification(isGroupChat, hasFilesUnknown, metrics.DiscardedReasonInternalError)
 			continue
 		}
 
 		post, skippedFileAttachments, _ := ah.msgToPost(channel.Id, botUserID, msg, chat, []string{})
 
-		ah.plugin.metricsService.ObserveNotification(len(chat.Members) >= 3, len(post.FileIds) > 0)
+		hasFiles := len(post.FileIds) > 0
```

Comment:
Technically, this should be the same for all users, but we only know it after we process the message and upload files on a per user basis.